### ISSUE TITLE
More information about error in HandleServerError

### ISSUE THREAD

**Issue by acarl005** [sveltejs/kit#6774]
### Describe the problem

Currently, the `handleError` hook in the server hooks gets triggered on a 404 error. I'm a bit confused about:

1. why [`HandleServerError`](https://kit.svelte.dev/docs/types#sveltejs-kit-handleservererror) needs to handle 404s. 
2. how [`HandleServerError`](https://kit.svelte.dev/docs/types#sveltejs-kit-handleservererror) can reliably tell the true "source" of an error coming from Kit, such as a 404.

## 1. Why Does HandleServerError need to handle 404?

Suppose i have a routeID like `/blogs/[slug]`, and the user requests `/blogs/nonexistent`. What makes sense to me is that in my `/src/routes/blogs/[slug]/+page.server.js`, I determine that the slug "nonexistent" does not exist and then I do something like this....

```js
import { error } from "@sveltejs/kit"

export function load() {
    throw error(404, "not blog with that slug")
}
```

`HandleServerError` is NOT triggered b/c this is an "expected error". This is imo the intuitive outcome because a 404 is the "client's fault", it is expected, and  not what `HandleServerError` is meant to handle, at least according to my understanding of [this section](https://kit.svelte.dev/docs/hooks#shared-hooks-handleerror)

> This function is not called for expected errors

**However**, if the user requests some route which doesn't have a directory in `routes/` at all, like `/totallyhasnohandler`, then a capital "E" Error is thrown by the framework, and `HandleServerError` *is* triggered. That is surprising to me. I don't want 404's to get included with the rest of my error-monitoring infra, which brings me to my next point....

## 2. How to reliably know why Kit threw this error and allow my `handleError` to react appropriately.

I want to divert things like 404s *away* from where the rest of the truly unexpected errors are going. Currently my hook looks somethign like this

```js
export function handleError({ error }) {
  if (error instanceof Error && error.message.includes("Not found:")) {
    return { message: "Not found" }
  }
  logToMoreIntenseErrorMonitoring(error)
  return { message: "internal error" }
}
```

Interestingly, Kit still knows to respond with a 404 to the request, but that state isn't getting passed to my hook, so I can't react to that. As a workaround, I am matching against the error message string which is getting me frowns from teammates. It just seems like there should be a more reliable way to get the source of this error. 

### Describe the proposed solution

404s either wouldn't trigger the error hook, or maybe 404s and other errors thrown by Kit would be more differentiated.... maybe the hook can have an option status code parameter if it has that? Maybe instead of the exception being a vanilla `Error` instance, it would be an `HttpError` instance which has a status?

### Alternatives considered

Matching against error message strings when I see them in my logs and I want to do different logic on them

### Importance

nice to have

### Additional Information

I have a little repro of this phenomenon: https://stackblitz.com/edit/sveltejs-kit-template-default-7jzfsn?file=src%2Fhooks.server.js&terminal=dev

I edited the nav to link to a non-existent route just for the purposes of triggering this case, but this is not indicative of a realistic scenario. In that case we would want to catch the broken link. But in the real world, clients and bots request all kinds of random URLs that trigger 404s.

**Comment by Conduitry** [comment]
If you want more control over 404s in particular, you can create a `[...catchAll]/+page.js` route with a `load` that does whatever sort of handling you want to do and then does `throw error(404, 'Not found')`. This gives you control over which layout(s) the error page is rendered inside, and also lets you handle it yourself knowing what it is, rather than having it lumped into `handleError`.

Is it just 404s you're concerned about? Would documenting this catchall route pattern somewhere help?

**Comment by acarl005** [comment]
Leaving this here in case other find this issue. After #6897 you can reliably detect a 404 error if `event.routeId === null`. So, my server hook `handleError` now looks something like this:

```js
export const handleError = ({ error, event }) => {
  if (event.routeId === null) {
    return {
      message: `Page ${event.url.pathname} not found`
    }
  }
  // ...
}
```
